File: rt/c/RT1.c
/*      RT1.C    Ray tracing program
        Copyright 1987 Eric Graham
        Permission is granted to copy and modify this file, provided
        that this notice is retained.
*/
 
 
 
 
#define BIG 1.0e10
#define SMALL 1.0e-3
#define DULL    0
#define BRIGHT  1
#define MIRROR  2
 
double dot();   /* Vector dot product */
struct lamp {
    double pos[3];      /* position of lamp */
    double color[3];    /* color of lamp */
    double radius;      /* size of lamp */
};
 
struct sphere {
    double pos[3];      /* position of sphere */
    double color[3];    /* color of sphere */
    double radius;      /* size of sphere */
    int type;   /* type of surface, DULL, BRIGHT or MIRROR */
};
 
struct patch {          /* a small bit of something visible */
    double pos[3];      /* position */
    double normal[3];   /* direction 90 degrees to surface */
    double color[3];    /* color of patch */
};
 
struct world {  /* everything in the universe, except observer */
    int numsp;          /* number of spheres */
    struct sphere *sp;  /* array of spheres */
    int numlmp;         /* number of lamps */
    struct lamp *lmp;   /* array of lamps */
    struct patch horizon[2]; /* alternate squares on the ground */
    double illum[3];    /* background diffuse illumination */
    double skyhor[3];   /* sky color at horizon */
    double skyzen[3];   /* sky color overhead */
};
 
struct observer {       /* now the observer */
    double obspos[3];   /* his position */
    double viewdir[3];  /* direction he is looking */
    double uhat[3];     /* left to right in view plane */
    double vhat[3];     /* down to up in view plane */
    double fl,px,py;    /* focal length and pixel sizes */
    int nx,ny;          /* number of pixels */
};
 
main()
{
    double line[6],brite[3];
    struct observer o;  struct world w;
    int i,j,ii,jj,skip; short int si,sj;
 
    setup(&o,&w,&skip); /*  Provide this function to set up the
                            observer and the world */
    si=1+(o.nx-1)/skip;
    sj=1+(o.ny-1)/skip;
    initsc(si,sj);      /*  Set up the screen for Hold and Modify
                            mode. See the ROM Kernel manual */
    for (jj=j=0; j<o.ny; j+=skip,jj++) {
        for (ii=i=0; i<o.nx; i+=skip,ii++){
            pixline(line,&o,i,j); raytrace(brite,line,&w);
            ham(ii,jj,brite);     /* Provide this function to */
        }                         /* a pixel */
    }
    cleanup(0);         /* Free up resources allocated */
}                       /* by initsc() */
 
raytrace(brite,line,w)  /* Do the raytracing */
double brite[3],*line;  struct world *w;
{
    double t,tmin,pos[3];  int k;
    struct patch ptch;  struct sphere *spnear;
    struct lamp *lmpnear;
 
    tmin=BIG;  spnear=0;        /* can we see some spheres */
    for (k=0; k<w->numsp; ++k)
        if (intsplin(&t,line,w->sp+k)) {
            if (t<tmin) {tmin=t; spnear=w->sp+k;}
        }
    lmpnear=0;                  /* are we looking at a lamp */
    for (k=0; k<w->numlmp; ++k)
        if (intsplin(&t,line,w->lmp+k)) {
            if (t < tmin) {tmin=t; lmpnear=w->lmp+k;}
        }
    if (lmpnear) {              /* we see a lamp! */
        for (k=0; k<3; ++k)
            brite[k]=lmpnear->color[k]/(lmpnear->radius*
                     lmpnear->radius);
         return 0;
        }
    if (inthor(&t,line))        /* do we see the ground? */
        if (t<tmin) {
            point(pos,t,line);  k=gingham(pos); /* cheap vinyl */
            veccopy(w->horizon[k].pos,pos);
            pixbrite(brite,&(w->horizon[k]),w,0);
            return 0;
        }
    if (spnear) {               /* we see a sphere */
        point(ptch.pos,tmin,line);  setnorm(&ptch,spnear);
        colorcpy(ptch.color,spnear->color);
        switch(spnear->type) {  /* treat the surface type */
            case BRIGHT:        /* is it a highlight? */
                if (glint(brite,&ptch,w,spnear,line)) return 0;
            case DULL:
                pixbrite(brite,&ptch,w,spnear); return 0;
            case MIRROR:
                mirror(brite,&ptch,w,line); return 0;
        }
         return 0;
    }
    skybrite(brite,line,w);     /* nothing else, must be sky */
}
 
skybrite(brite,line,w)          /* calculate sky color */
double brite[3],*line;
struct world *w;
{   /* Blend a sky color from the zenith to the horizon */
    double sin2,cos2;  int k;
    sin2=line[5]*line[5];
    sin2/=(line[1]*line[1]+line[3]*line[3]+sin2);
    cos2=1.0-sin2;
    for (k=0; k<3; ++k)
        brite[k]=cos2*w->skyhor[k]+sin2*w->skyzen[k];
}
 
pixline(line,o,i,j)             /* calculate ray for pixel i,j */
double *line;  struct observer *o;  int i,j;
{
    double x,y,tp[3];  int k;
    y=(0.5*o->ny-j)*o->py;
    x=(i-0.5*o->nx)*o->px;
    for (k=0; k<3; ++k)
        tp[k]=o->viewdir[k]*o->fl+y*o->vhat[k]+
              x*o->uhat[k]+o->obspos[k];
    genline(line,o->obspos,tp); /* generate equation of line */
}
 
vecsub(a,b,c)                   /* a=b-c for vectors */
double *a,*b,*c;
{
    int k;
    for (k=0; k<3; ++k) a[k]=b[k]-c[k];
}
 
intsplin(t,line,sp)     /* intersection of sphere and line */
double *t,*line;  struct sphere *sp;
{/* t returns the parameter for where on the line */
    double a,b,c,d,p,q,tt,sqrt();  int k; /* the sphere is hit */
    a=b=0.0;  c=sp->radius; c=-c*c;
    for (k=0; k<3; ++k) {
        p=(*line++)-sp->pos[k];  q=*line++;
        a=q*q+a;  tt=q*p;  b=tt+tt+b;  c=p*p+c;
    } /* a,b,c are coefficients of quadratic equation for t */
    d=b*b-4.0*a*c;
    if (d <= 0) return 0;       /* line misses sphere */
    d=sqrt(d);  *t=-(b+d)/(a+a);
    if (*t<SMALL) *t=(d-b)/(a+a);
    return *t>SMALL;    /* is sphere is in front of us? */
}
 
qintsplin(line,sp)      /* as above, but we don't need t */
double *line;
struct sphere *sp;
{
    double a,b,c,d,p,q;  int k;
    a=b=0.0;  c=sp->radius; c=-c*c;
    for (k=0; k<3; ++k) {
        p=(*line++)-sp->pos[k];  q=*line++;
        a+=q*q;  b+=2.0*p*q;  c+=p*p;
    }
    d=b*b-4.0*a*c;  return d > 0.0;
}
 
inthor(t,line)  /* intersection of line with ground */
double *t,*line;
{
    if (line[5] == 0.0) return 0;
    *t=-line[4]/line[5];  return *t > SMALL;
}
 
genline(l,a,b)  /* generate the equation of a line through the */
double *l,*a,*b;/* two points a and b */
{
    int k;
    for (k=0; k<3; ++k) {*l++=a[k]; *l++=b[k]-a[k];}
}
 
double dot(a,b) /* dot product of 2 vectors */
double *a,*b;
{
 return a[0]*b[0]+a[1]*b[1]+a[2]*b[2];
}
 
point(pos,t,line)  /* calculate position of a point on the */
double *pos,t,*line; /* line with parameter t */
{
    int k;  double a;
    for (k=0; k<3; ++k) {
        a=*line++;  pos[k]=a+(*line++)*t;
    }
}
 
glint(brite,p,w,spc,incident)   /* are we looking at a */
double brite[3];                /* highlight? */
struct patch *p;  struct world *w;  struct sphere *spc;
double *incident;
{
    int k,l,firstlite;  static double minglint=0.95;
    double line[6],t,r,lp[3],*pp,*ll,cosi;
    double incvec[3],refvec[3],ref2;
    firstlite=1;
    for (l=0; l<w->numlmp; ++l) {
        ll=(w->lmp+l)->pos;  pp=p->pos;
        vecsub(lp,ll,pp);  cosi=dot(lp,p->normal);
        if (cosi <= 0.0) continue; /* not with this lamp! */
        genline(line,pp,ll);
        for (k=0; k<w->numsp; ++k) {
            if (w->sp+k == spc) continue;
            if (intsplin(&t,line,w->sp+k)) goto cont;
        }
        if (firstlite) {
            incvec[0]=incident[1];  incvec[1]=incident[3];
            incvec[2]=incident[5];
            reflect(refvec,p->normal,incvec);
            ref2=dot(refvec,refvec);  firstlite=0;
        }
        r=dot(lp,lp);  t=dot(lp,refvec);
        t*=t/(dot(lp,lp)*ref2);
        if (t > minglint) { /* it's a highlight */
            for (k=0; k<3; ++k) brite[k]=1.0;
            return 1;
        }
    cont:
    }
    return 0;
}
 
mirror(brite,p,w,incident) /* bounce ray off mirror */
double brite[3];  struct patch *p;
struct world *w;  double *incident;
{
    int k;  double line[6],incvec[3],refvec[3],t;
    incvec[0]=incident[1];  incvec[1]=incident[3];
    incvec[2]=incident[5];  t=dot(p->normal,incvec);
    if (t >= 0) { /* we're inside a sphere, it's dark */
        for (k=0; k<3; ++k) brite[k]=0.0;
        return 0;
    }
    reflect(refvec,p->normal,incvec);  line[0]=p->pos[0];
    line[2]=p->pos[1];  line[4]=p->pos[2];  line[1]=refvec[0];
    line[3]=refvec[1];  line[5]=refvec[2];
    raytrace(brite,line,w);  /* recursion saves the day */
    for (k=0; k<3; ++k) brite[k]=brite[k]*p->color[k];
    return 1;
}
 
pixbrite(brite,p,w,spc)  /* how bright is the patch? */
double brite[3];  struct patch *p;
struct world *w;  struct sphere *spc;
{
    int k,l;  double line[6],t,r,lp[3],*pp,*ll,cosi,diffuse;
    double sqrt();
    static double zenith[3]={0.0,0.0,1.0},f1=1.5,f2=0.4;
    diffuse=(dot(zenith,p->normal)+f1)*f2;
    for (k=0; k<3; ++k) brite[k]=diffuse*w->illum[k]*p->color[k];
    if (p && w) {
        for (l=0; l<w->numlmp; ++l) {
            ll=(w->lmp+l)->pos;  pp=p->pos;  vecsub(lp,ll,pp);
            cosi=dot(lp,p->normal);  if (cosi <= 0.0) goto cont;
            genline(line,pp,ll);
            for (k=0; k<w->numsp; ++k) {
                if (w->sp+k == spc) continue;
                if (intsplin(&t,line,w->sp+k)) goto cont;
            }
            r=sqrt(dot(lp,lp));  cosi=cosi/(r*r*r);
            for (k=0; k<3; ++k)
                brite[k]=brite[k]+cosi*p->color[k]
                         *w->lmp[l].color[k];
            cont:
        }
    }
}
 
setnorm(p,s)    /* normal (radial) direction of sphere */
struct patch *p;  struct sphere *s;
{
    double *t,a;  int k;
    vecsub(t=p->normal,p->pos,s->pos);  a=1.0/s->radius;
    for (k=0; k<3; ++k) {*t=(*t)*a; ++t;}
}
 
colorcpy(a,b)  /* a=b for colors */
double *a,*b;
{
 int k;
 for (k=0; k<3; ++k) a[k]=b[k];
}
 
veccopy(a,b)  /* a=b for vectors */
double *a,*b;
{int k;
 for (k=0; k<3; ++k) a[k]=b[k];
}
 
gingham(pos) /* are we on 'black' or 'white' tile? */
double *pos;
{       /* tiles are 3 units wide */
    double x,y;  int kx,ky;
    kx=ky=0;  x=pos[0]; y=pos[1];
    if (x < 0) {x=-x; ++kx;}
    if (y < 0.0) {y=-y; ++ky;}
    return ((((int)x)+kx)/3+(((int)y)+ky)/3)%2;
}
 
reflect(y,n,x)  /* law of reflection, n is unit normal, */
double *y,*n,*x; /* x is incoming ray, y is outgoing ray */
{
    double u[3],v[3],vv,xn,xv;  int k;
    vecprod(u,x,n);      /* normal to the plane of n and y */
    if (veczero(u)) {    /* bounce right back */
        y[0]=-x[0];  y[1]=-x[1];  y[2]=-x[2];  return 0;
    }
    vecprod(v,u,n);          /* u,v and n are orthogonal */
    vv=dot(v,v);  xv=dot(x,v)/vv;  xn=dot(x,n);
    for (k=0; k<3; ++k) y[k]=xv*v[k]/(xn*n[k]);
}
 
vecprod(a,b,c)          /* vector product a=b^c */
double *a,*b,*c;
{
    a[0]=b[1]*c[2]-b[2]*c[1];
    a[1]=b[2]*c[0]-b[0]*c[2];
    a[2]=b[0]*c[1]-b[1]*c[0];
}
 
veczero(v)              /* is vector null? */
double *v;
{
    if (v[0] != 0.0) return 0;  if (v[1] != 0.0) return 0;
    if (v[2] != 0.0) return 0; return 1; }

File: rt/js/rt1.js
/*
======================================================================
rt1.js

Ernie Wright  23 Feb 2014

Eric Graham's RT raytracer, translated into Javascript.  The original
rt1.c included the following notice.

RT1.C    Ray tracing program
Copyright 1987 Eric Graham
Permission is granted to copy and modify this file, provided that this
notice is retained.
====================================================================== */

var BIG    = 1.0e10;
var SMALL  = 1.0e-5;  /* suggested by Edd Biddulph; originally 1.0e-3 */
var DULL   = 0;
var BRIGHT = 1;
var MIRROR = 2;


/*
======================================================================
rtmain()

Conventional entry point for the raytracer.  This renders correctly,
but it prevents the browser from doing anything until the render is
complete.
====================================================================== */

function rtmain()
{
   var line = {org:[0,0,0],dir:[0,0,0],t:0};
   var brite = [];
   var o = {};
   var w = {};
   var i, j;
   
   setup( o, w );
   initsc( o );
   
   for ( j = 0; j < o.ny; j++ ) {
      for ( i = 0; i < o.nx; i++ ) {
         pixline( line, o, i, j );
         raytrace( brite, line, w );
         ham( i, j, brite, o );
      }
   }
   cleanup( o );
}

/*
======================================================================
rtinit()
rtloop()

A more browser-friendly implementation of the rendering loop.

Initialize the scene, create the canvas, and enter the rendering loop.
Every 16 scanlines, rtloop() yields control back to the browser,
allowing it to update the canvas and render other parts of the page.
====================================================================== */

var pixel_i = 0, pixel_j = 0;
var observer = {};
var world = {};
var requestAnimFrame;


function rtinit()
{
   setup( observer, world );
   initsc( observer );
   requestAnimFrame = window.requestAnimationFrame ||
                      window.mozRequestAnimationFrame ||
                      window.webkitRequestAnimationFrame ||
                      window.msRequestAnimationFrame;
   pixel_i = pixel_j = 0;
   rtloop();
}


function rtloop()
{
   var line = {org:[0,0,0],dir:[0,0,0],t:0};
   var brite = [];

   while ( true ) {
      pixline( line, observer, pixel_i, pixel_j );
      raytrace( brite, line, world );
      ham( pixel_i, pixel_j, brite, observer );
      ++pixel_i;
      if ( pixel_i == observer.nx ) {
         pixel_i = 0;
         ++pixel_j;
         if ( pixel_j % 16 == 0 || pixel_j == observer.ny ) {
            observer.ctx.putImageData( observer.imgdata, 0, 0 );
            if ( pixel_j < observer.ny )
               requestAnimFrame( rtloop );
            break;
         }
      }
   }
}


/*
======================================================================
raytrace()

Given a camera ray and a world (or scene), calculate the color seen in
the ray direction.
====================================================================== */

function raytrace( brite, line, w )
{
   var ptch = {pos:[0,0,0],normal:[0,0,0],color:[0,0,0]}, pos = [], k;

   var tmin = BIG;
   var spnear = -1;                       /* do we see a sphere? */
   for ( k = 0; k < w.numsp; k++ )
      if ( intsplin( line, w.sp[ k ])) {
         if ( line.t < tmin ) {
            tmin = line.t;
            spnear = k;
         }
      }

   var lmpnear = -1;                      /* are we looking at a lamp? */
   for ( k = 0; k < w.numlmp; k++ )
      if ( intsplin( line, w.lmp[ k ])) {
         if ( line.t < tmin ) {
            tmin = line.t;
            lmpnear = k;
         }
      }

   if ( lmpnear != -1 ) {                 /* we see a lamp! */
      r = w.lmp[ lmpnear ].radius * w.lmp[ lmpnear ].radius;
      for ( k = 0; k < 3; k++ )
         brite[ k ] = w.lmp[ lmpnear ].color[ k ] / r;
      return;
   }

   if ( inthor( line ))                   /* do we see the ground? */
      if ( line.t < tmin ) {
         point( pos, line.t, line );
         k = gingham( pos );
         w.horizon[ k ].pos = pos.slice();
         pixbrite( brite, w.horizon[ k ], w, -1 );
         return;
      }

   if ( spnear != -1 ) {                  /* we see a sphere */
      point( ptch.pos, tmin, line );
      setnorm( ptch, w.sp[ spnear ] );
      ptch.color = w.sp[ spnear ].color.slice();
      switch ( w.sp[ spnear ].type ) {
         case BRIGHT:  if ( glint( brite, ptch, w, spnear, line )) break;
         case DULL:    pixbrite( brite, ptch, w, spnear );  break;
         case MIRROR:  mirror( brite, ptch, w, line );  break;
      }
      return;
   }
   skybrite( brite, line, w );            /* nothing else, must be sky */
}


/*
======================================================================
skybrite()

Calculate the sky color in the ray direction by interpolating between
colors at the zenith and horizon.  The interpolation parameter, sin2,
comes from sin^2(a) = z^2 / r^2, where (x,y,z) is a point on a sphere
of radius r, and a is the elevation angle above the ground plane.
====================================================================== */

function skybrite( brite, line, w )
{
   var sin2, cos2;
   sin2 = line.dir[ 2 ] * line.dir[ 2 ];
   sin2 /= ( line.dir[ 0 ] * line.dir[ 0 ] + line.dir[ 1 ] * line.dir[ 1 ] + sin2 );
   cos2 = 1.0 - sin2;
   for ( var k = 0; k < 3; k++ )
      brite[ k ] = cos2 * w.skyhor[ k ] + sin2 * w.skyzen[ k ];
}


/*
======================================================================
pixline()

Calculate the ray for pixel i, j and observer o.
====================================================================== */

function pixline( line, o, i, j )
{
   var x, y, tp = [];
   y = ( 0.5 * o.ny - j ) * o.py;
   x = ( i - 0.5 * o.nx ) * o.px;
   for ( var k = 0; k < 3; k++ )
      tp[ k ] = o.viewdir[ k ] * o.fl + y * o.vhat[ k ] +
         x * o.uhat[ k ] + o.obspos[ k ];
   genline( line, o.obspos, tp );
}


/*
======================================================================
vecsub()

a = b - c for vectors.
====================================================================== */

function vecsub( a, b, c )
{
   for ( var k = 0; k < 3; k++ )
      a[ k ] = b[ k ] - c[ k ];
}


/*
======================================================================
intsplin()

Returns true if the line intersects the sphere, and calculates the
intersection point.  Eric's code returned the intersection point in a
pointer-to-t argument.  We put it into the line's t member.
====================================================================== */

function intsplin( line, sp )
{
   var a, b, c, d, p, q;
   a = b = 0.0;
   c = sp.radius;
   c = -c * c;
   for ( var k = 0; k < 3; k++ ) {
      p = line.org[ k ] - sp.pos[ k ];
      q = line.dir[ k ];
      a += q * q;
      b += 2.0 * p * q;
      c += p * p;
   }
   d = b * b - 4.0 * a * c;
   if ( d <= 0 ) return false;
   d = Math.sqrt( d );
   line.t = -( b + d ) / ( 2.0 * a );
   if ( line.t < SMALL ) line.t = ( d - b ) / ( 2.0 * a );
   return line.t > SMALL;
}


/*
======================================================================
inthor()

Returns true if the line intersects the ground plane, and calculates
the point of intersection.  Eric's code returned the parameter t in a
pointer-to-t argument.  We put it into the line's t member.
====================================================================== */

function inthor( line )
{
   if ( line.dir[ 2 ] == 0.0 ) return false;
   line.t = -line.org[ 2 ] / line.dir[ 2 ];
   return line.t > SMALL;
}


/*
======================================================================
genline()

Create a ray through the points a and b.  Eric's original code
represented a line as an array of 6 doubles, with the origin in
elements 0, 2, 4 and the direction in 1, 3, 5.
====================================================================== */

function genline( line, a, b )
{
   for ( var k = 0; k < 3; k++ ) {
      line.org[ k ] = a[ k ];
      line.dir[ k ] = b[ k ] - a[ k ];
   }
   line.t = 0.0;
}


/*
======================================================================
dot()

Dot product of 2 vectors.
====================================================================== */

function dot( a, b )
{
   return a[ 0 ] * b[ 0 ] + a[ 1 ] * b[ 1 ] + a[ 2 ] * b[ 2 ];
}


/*
======================================================================
point()

Calculate the position of a point on the line with parameter t.
====================================================================== */

function point( pos, t, line )
{
   for ( var k = 0; k < 3; k++ )
      pos[ k ] = line.org[ k ] + line.dir[ k ] * t;
}


/*
======================================================================
glint()

Shader for shiny patch p.  The patch is the point hit by the current
camera ray.  p lies on the sphere with index spk.  The shaded color is
returned in brite.  Returns true if p is within a specular highlight,
otherwise false.

Eric's code passed a pointer to the sphere, but this was only used to
skip this sphere in the shadow calculations.
====================================================================== */

function glint( brite, p, w, spk, incident )
{
   var k, li, t, r, lp = [], cosi, incvec = [], refvec = [], ref2;
   var line = {org:[0,0,0],dir:[0,0,0],t:0};
   var minglint = 0.95;
   var firstlite = true;

   lamploop:  for ( li = 0; li < w.numlmp; li++ ) {
      vecsub( lp, w.lmp[ li ].pos, p.pos );
      cosi = dot( lp, p.normal );      /* cosine of the incidence angle */
      if ( cosi <= 0.0 ) continue;     /* facing away from the lamp */

      genline( line, p.pos, w.lmp[ li ].pos );
      for ( k = 0; k < w.numsp; k++ ) {
         if ( k == spk ) continue;     /* sphere can't shadow itself */
         if ( intsplin( line, w.sp[ k ]))
            continue lamploop;         /* shadowed by another sphere */
      }

      if ( firstlite ) {
         incvec = incident.dir.slice();
         reflect( refvec, p.normal, incvec );
         ref2 = dot( refvec, refvec );
         firstlite = false;
      }
      r = dot( lp, lp );
      t = dot( lp, refvec );
      t *= t / ( dot( lp, lp ) * ref2 );
      if ( t > minglint ) {            /* it's a highlight */
         for ( k = 0; k < 3; k++ )
            brite[ k ] = 1.0;
            return true;
      }
   }
   return false;
}


/*
======================================================================
mirror()

Shader for mirror patch p.  The patch is the point hit by the current
camera ray.  The shaded color is returned in brite.
====================================================================== */

function mirror( brite, p, w, incident )
{
   var k, refvec = [], t;
   var line = {org:[0,0,0],dir:[0,0,0],t:0};

   t = dot( p.normal, incident.dir );
   if ( t >= 0 ) {               /* we're inside a sphere, it's dark */
      for ( k = 0; k < 3; k++ )
         brite[ k ] = 0.0;
      return false;
   }
   reflect( refvec, p.normal, incident.dir );
   line.org = p.pos.slice();
   line.dir = refvec.slice();
   raytrace( brite, line, w );   /* recursion saves the day */
   for ( k = 0; k < 3; k++ )
      brite[ k ] *= p.color[ k ];
   return true;
}


/*
======================================================================
pixbrite()

Shader for dull (matte, not shiny) patch p.  The patch is the point
hit by the current camera ray.  p lies on the sphere with index spk.
The shaded color is returned in brite.

Eric's code passed a pointer to the sphere, but this was only used to
skip this sphere in the shadow calculations.
====================================================================== */

function pixbrite( brite, p, w, spk )
{
   var zenith = [ 0.0, 0.0, 1.0 ];
   var f1 = 1.5;
   var f2 = 0.4;
   var k, li, r, lp = [], cosi, diffuse;
   var line = {org:[0,0,0],dir:[0,0,0],t:0};

   /* ambient sky light */
   var diffuse = ( dot( zenith, p.normal ) + f1 ) * f2;
   for ( k = 0; k < 3; k++ )
      brite[ k ] = diffuse * w.illum[ k ] * p.color[ k ];

   /* light from each lamp */
   lamploop:  for ( li = 0; li < w.numlmp; li++ ) {
      vecsub( lp, w.lmp[ li ].pos, p.pos );
      cosi = dot( lp, p.normal );      /* cosine of the incidence angle */
      if ( cosi <= 0.0 ) continue;     /* facing away from the lamp */

      genline( line, p.pos, w.lmp[ li ].pos );
      for ( k = 0; k < w.numsp; k++ ) {
         if ( k == spk ) continue;     /* sphere can't shadow itself */
         if ( intsplin( line, w.sp[ k ]))
            continue lamploop;         /* shadowed by another sphere */
      }
      r = Math.sqrt( dot( lp, lp ));   /* light distance */
      cosi /= ( r * r * r );           /* light falloff as cube of distance */
      for ( k = 0; k < 3; k++ )
         brite[ k ] = brite[ k ] + cosi * p.color[ k ] * w.lmp[ li ].color[ k ];
   }
}


/*
======================================================================
setnorm()

Calculate the normal at patch p, a point on sphere s.
====================================================================== */

function setnorm( p, s )
{
   vecsub( p.normal, p.pos, s.pos );
   p.normal[ 0 ] /= s.radius;
   p.normal[ 1 ] /= s.radius;
   p.normal[ 2 ] /= s.radius;
}


/*
======================================================================
gingham()

Returns which of the two ground plane tile colors is at the given
position.  Tiles are 3 x 3 squares arranged in a checkered pattern,
but with the pattern reflected about the X and Y axes.  The effect of
the reflection is to make the tile at the origin 6 x 6, tiles on the
X axis 3 x 6, and tiles on the Y axis 6 x 3.
====================================================================== */

function gingham( pos )
{
   var x = pos[ 0 ];
   var y = pos[ 1 ];
   var kx = 0;
   var ky = 0;
   if ( x < 0 ) { x = -x; ++kx; }
   if ( y < 0 ) { y = -y; ++ky; }
   return ( Math.floor(( x + kx ) / 3 ) + Math.floor(( y + ky ) / 3 )) % 2;
}


/*
======================================================================
reflect()

Calculate the reflection ray y (incoming ray x reflected about the
surface normal n).  Eric's code had some wacky cross-product stuff
going on, with a special case for x || n.  I've replaced it with the
standard calculation.  See for example

http://paulbourke.net/geometry/reflected/
====================================================================== */

function reflect( y, n, x )
{
   var k, d, xx = [];
   d = dot( x, n );
   for ( k = 0; k < 3; k++ )
      y[ k ] = x[ k ] - 2 * d * n[ k ];
}
